tfhe/src/integer/server_key/radix/scalar_sub.rs

```diff
@@ -97,37 +125,23 @@ impl ServerKey {
     ///
     /// assert_eq!(true, res);
     /// ```
-    pub fn is_scalar_sub_possible<PBSOrder: PBSOrderMarker>(
+    pub fn is_scalar_sub_possible<PBSOrder, T>(
         &self,
         ct: &RadixCiphertext<PBSOrder>,
-        scalar: u64,
-    ) -> bool {
-        //Bits of message put to 1
-        let mask = (self.key.message_modulus.0 - 1) as u64;
-
-        let modulus = self
-            .key
-            .message_modulus
-            .0
-            .saturating_pow(ct.blocks.len() as u32) as u64;
-
-        let neg_scalar = scalar.wrapping_neg() % modulus;
-
-        let mut power = 1_u64;
-
-        for ct_i in ct.blocks.iter() {
-            let mut decomp = neg_scalar & (mask * power);
-            decomp /= power;
-
-            if !self.key.is_scalar_add_possible(ct_i, decomp as u8) {
-                return false;
-            }
-
-            //modulus to the power i
-            let Some(new_power) = power.checked_mul(self.key.message_modulus.0 as u64) else {break};
-            power = new_power;
-        }
-        true
+        scalar: T,
+    ) -> bool
+    where
+        PBSOrder: PBSOrderMarker,
+        T: UnsignedInteger + DecomposableInto<u8>,
+    {
+        let decomposer = self.create_negated_block_decomposer(scalar);
+        ct.blocks
+            .iter()
+            .zip(decomposer)
+            .all(|(ciphertext_block, scalar_block)| {
+                self.key
+                    .is_scalar_add_possible(ciphertext_block, scalar_block)
```

Comment:
perhaps here a small comment explaining we are using 2's complement and so the add is the operation we want (?)